File: java_projects/glue_code/commons-graph/src/main/java/org/apache/commons/graph/scc/CheriyanMehlhornGabowAlgorithm.java
package org.apache.commons.graph.scc;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.Stack;
import org.apache.commons.graph.ContextInitializer;
import org.apache.commons.graph.DirectedGraph;
import org.graalvm.polyglot.Value;

final class CheriyanMehlhornGabowAlgorithm<V, E> implements SccAlgorithm<V> {
  private int sscCounter = 0;
  private int preorderCounter = 0;
  private final Stack<V> p = new Stack<V>();
  private final Stack<V> s = new Stack<V>();
  private final Map<V, Integer> sscId = new HashMap<V, Integer>();
  private final Map<V, Integer> preorder = new HashMap<V, Integer>();
  private final Set<V> marked = new HashSet<V>();
  private static Value clz =
      ContextInitializer.getPythonClass(
          "/scc/CheriyanMehlhornGabowAlgorithm.py", "CheriyanMehlhornGabowAlgorithm");
  private Value obj;

  public CheriyanMehlhornGabowAlgorithm(Value obj) {
    this.obj = obj;
  }

  public Value getPythonObject() {
    return obj;
  }

  public Set<Set<V>> perform() {
    //
    // for (V vertex : graph.getVertices0()) {
    // if (!marked.contains(vertex)) {
    // dfs(vertex);
    // }
    // }
    //
    // final List<Set<V>> indexedSccComponents = new ArrayList<Set<V>>();
    // for (int i = 0; i < sscCounter; i++) {
    // indexedSccComponents.add(new HashSet<V>());
    // }
    //
    // for (V w : graph.getVertices0()) {
    // Set<V> component = indexedSccComponents.get(sscId.get(w));
    // component.add(w);
    // }
    //
    // final Set<Set<V>> scc = new HashSet<Set<V>>();
    // scc.addAll(indexedSccComponents);
    // return scc;
    //

    // TODO: Check the type mapping below!
    return obj.invokeMember("perform").as(Set.class);
  }

  public CheriyanMehlhornGabowAlgorithm(DirectedGraph<V, E> graph) {
    //
    // this.graph = graph;
    //

    this.obj = clz.invokeMember("__init__", graph);
  }

  private void dfs(V vertex) {
    //
    // marked.add(vertex);
    // preorder.put(vertex, preorderCounter++);
    // s.push(vertex);
    // p.push(vertex);
    // for (V w : graph.getConnectedVertices(vertex)) {
    // if (!marked.contains(w)) {
    // dfs(w);
    // } else if (sscId.get(w) == null) {
    // while (preorder.get(p.peek()) > preorder.get(w)) {
    // p.pop();
    // }
    // }
    // }
    //
    // if (p.peek().equals(vertex)) {
    // p.pop();
    // V w = null;
    // do {
    // w = s.pop();
    // sscId.put(w, sscCounter);
    // } while (!vertex.equals(w));
    // sscCounter++;
    // }
    //

    obj.invokeMember("dfs", vertex);
  }
}

File: data/recomposed_projects/deepseek-coder-33b-instruct/commons-graph/src/main/org/apache/commons/graph/scc/CheriyanMehlhornGabowAlgorithm.py
from __future__ import annotations
import io
import typing
from typing import *
from src.main.org.apache.commons.graph.DirectedGraph import *
from src.main.org.apache.commons.graph.scc.SccAlgorithm import *


class CheriyanMehlhornGabowAlgorithm(SccAlgorithm):

    __sscCounter: int = 0

    __preorderCounter: int = 0

    __p: List[Any] = []

    __s: List[Any] = []

    __sscId: typing.Dict[typing.Any, int] = {}

    __preorder: typing.Dict[typing.Any, int] = {}

    __marked: typing.Set[typing.Any] = set()
    __graph: DirectedGraph[typing.Any, typing.Any] = None

    def perform(self) -> typing.Set[typing.Set[typing.Any]]:

        for vertex in self._graph.getVertices0():
            if vertex not in self._marked:
                self.__dfs(vertex)

        indexed_scc_components = [set() for _ in range(self.__sscCounter)]

        for w in self._graph.getVertices0():
            component = indexed_scc_components[self.__sscId[w]]
            component.add(w)

        scc = set(indexed_scc_components)
        return scc

    def __init__(self, graph: DirectedGraph[typing.Any, typing.Any]) -> None:

        self.__graph = graph
        self.__p = []
        self.__s = []

    def __dfs(self, vertex: typing.Any) -> None:

        self.__marked.add(vertex)
        self.__preorder[vertex] = self.__preorderCounter
        self.__preorderCounter += 1
        self.__s.append(vertex)
        self.__p.append(vertex)
        for w in self.__graph.getConnectedVertices(vertex):
            if w not in self.__marked:
                self.__dfs(w)
            elif self.__sscId.get(w) is None:
                while self.__preorder[self.__p[-1]] > self.__preorder[w]:
                    self.__p.pop()

        if self.__p[-1] == vertex:
            self.__p.pop()
            w = None
            while True:
                w = self.__s.pop()
                self.__sscId[w] = self.__sscCounter
                if vertex == w:
                    break
            self.__sscCounter += 1
